api/src/pcapi/core/users/api.py

```diff
@@ -833,52 +834,47 @@ def is_user_age_compatible_with_eligibility(user_age: int | None, eligibility: m
     filters = []
-    name_terms: list[str] = []
+    name_term: str = ""
 
-    if not terms:
+    if not term or "%" in term:
         return accounts.filter(False)
 
-    for term in terms:
-        if not term:
-            continue
+    term_filters: list[sa.sql.ColumnElement] = []
 
-        term_filters: list[sa.sql.ColumnElement] = []
-
-        # phone number
-        try:
-            parsed_phone_number = phone_number_utils.parse_phone_number(term)
-            term_as_phone_number = phone_number_utils.get_formatted_phone_number(parsed_phone_number)
-        except phone_validation_exceptions.InvalidPhoneNumber:
-            pass  # term can't be a phone number
-        else:
-            term_filters.append(models.User.phoneNumber == term_as_phone_number)  # type: ignore [arg-type]
-
-        # numeric
-        if term.isnumeric():
-            term_filters.append(models.User.id == int(term))
-
-        # email
-        sanitized_term = email_utils.sanitize_email(term)
-        if email_utils.is_valid_email(sanitized_term):
-            term_filters.append(models.User.email == sanitized_term)
-        elif email_utils.is_valid_email_domain(sanitized_term):
-            # search for all emails @domain.ext
-            term_filters.append(models.User.email.like(f"%{sanitized_term}"))
-
-        if not term_filters:
-            # terms which do not match a formatted column are searched by trigram distance in the names
-            term_filters.append(
-                sa.func.similarity(models.User.firstName, term) > settings.BACKOFFICE_SEARCH_SIMILARITY_MINIMAL_SCORE
-            )
+    # phone number
+    try:
+        parsed_phone_number = phone_number_utils.parse_phone_number(term)
+        term_as_phone_number = phone_number_utils.get_formatted_phone_number(parsed_phone_number)
+    except phone_validation_exceptions.InvalidPhoneNumber:
+        pass  # term can't be a phone number
+    else:
+        term_filters.append(models.User.phoneNumber == term_as_phone_number)  # type: ignore [arg-type]
+
+    # numeric
+    if term.isnumeric():
+        term_filters.append(models.User.id == int(term))
+
+    # email
+    sanitized_term = email_utils.sanitize_email(term)
+    if email_utils.is_valid_email(sanitized_term):
+        term_filters.append(models.User.email == sanitized_term)
+    elif email_utils.is_valid_email_domain(sanitized_term):
+        # search for all emails @domain.ext
+        term_filters.append(models.User.email.like(f"%{sanitized_term}"))
+
+    if not term_filters:
+        name_term = term
+        for name in name_term.split():
             term_filters.append(
-                sa.func.similarity(models.User.lastName, term) > settings.BACKOFFICE_SEARCH_SIMILARITY_MINIMAL_SCORE
+                sa.func.unaccent(sa.func.concat(models.User.firstName, " ", models.User.lastName)).ilike(
+                    f"%{clean_accents(name)}%"
+                )
```

Comment:
Donc on concatène le nom et le prénom et on fait un ILIKE dessus c'est ça ?
Avec un EXPLAIN, ça donne quoi par exemple ?

:arrow_right: Est-ce que tu as essayé quelque chose du style : 

```sql
[WHERE] ... 
    (user."firstName" ILIKE '<first_name>%'
    OR user."lastName" ILIKE '<last_name>%')
```

Si oui, qu'est-ce que ça a donné ?

Reply:
J'essaye avec EXPLAIN soon.
Je n'ai pas essayé avec first_name last_name, parce que ça implique soit de déterminer soit même ce qui est quoi (complexe quand on cherche Abdel Yves Akhim Thomas par exemple), soit d'avoir des champs précis dans l'interface de la recherche, ce qui n'est pas encore le cas.

Reply:
Je pose la question parce que si j'ai bien compris, le code va générer une requête du style : 

```sql
select
    id, email, "firstName", "lastName"
from
    public."user"
where
    CONCAT("firstName", ' ', "lastName") ilike '%tristan%';
```

Ce qui est **sensiblement** plus lent que quelque chose comme : 

```sql
select
    id, email, "firstName", "lastName"
from
    public."user"
where
    "firstName" ilike ('tristan%') or "lastName" ilike('tristan%')
```

Reply:
En supposant qu'on cherche avec nom prénom (parce que c'est le cas majoritaire de toutes les requêtes qui ne sont pas par mail/id/tel), et notamment un nom composé (e.g. comme dans staging, Diane Sou de Laroque).

On aurait une requête
```
select
    id, email, "firstName", "lastName"
from
    public."user"
where
    "firstName" ilike ('Diane%') or "lastName" ilike('Diane%')
    AND "firstName" ilike ('Sou%') or "lastName" ilike('Sou%')
    AND "firstName" ilike ('de%') or "lastName" ilike('de%')
    AND "firstName" ilike ('Laroque%') or "lastName" ilike('Laroque%')
```
Et là, on ne trouve pas le résultat, car on doit avoir lastName tel qu'il commence par Sou, de et Laroque.

Reply:
Dans la PR, quand on cherche par exemple "Diane Sou de Laroque", on va split ceci en `term`, et faire une suite de `AND concat(etc) ILIKE(%term%)`. L'idée est de trouver tous les users qui ont ces termes dans le nom. Et ça, ça vient du fait qu'un user qui se plaint va peut-être pas donner autant d'infos que ce qui est dans sa fiche user.

Supposons un "Venceslas Roullier de Passe de Culture", il va probablement dire au support qu'il s'appelle "Venceslas de Passe". Quelqu'un avec un deuxième prénom "français", e.g. un "Kazuko Louis Roullier" dira peut-être qu'il s'appelle "Louis Roullier". D'où les "%" avant ET après chaque terme.